### PR TITLE
[FLINK-35464] Fixes operator state backwards compatibility from CDC 3.0.x

### PR DESCRIPTION
**Pull request by yuxiqian**
This closes FLINK-35441 and FLINK-35464.

Flink CDC 3.1 changes how SchemaRegistry [de]serializes state data, which causes any checkpoint states saved with earlier version could not be restored in version 3.1.0. This PR adds serialization versioning for state payloads and ensures 3.0.x state could be successfully restored.

Unfortunately 3.1.0 introduces breaking changes without bumping serialization version, so this release will be excluded from state compatibility guarantee.

---

Running some real savepoint restore tests with https://github.com/yuxiqian/migration-test reveals that:

Before this change: (Snapshots compiled from `release-3.1` and `master`)

|  From \ To   | 3.0.0 | 3.0.1 | 3.1.0 | 3.1-SNAPSHOT | 3.2-SNAPSHOT |
| ------------ | ----- | ----- | ----- | ------------ | -------------|
| 3.0.0        | ❓    | ❓    | ❌     | ❌            | ❌           |
| 3.0.1        |       | ❓    | ❌     | ❌            | ❌           |
| 3.1.0        |       |       | ✅     | ✅            | ✅           |
| 3.1-SNAPSHOT |       |       |        | ✅            | ✅           |    
| 3.2-SNAPSHOT |       |       |        |               | ✅           |

> ✅ - Compatible, ❌ - Not compatible, ❓ - Target version doesn't support `--from-savepoint`

After this change: (Snapshots compiled from `FLINK-35464-BP-3.1` and `FLINK-35464`)

|  From \ To   | 3.0.0 | 3.0.1 | 3.1.0 | 3.1-SNAPSHOT | 3.2-SNAPSHOT |
| ------------ | ----- | ----- | ----- | ------------ | -------------|
| 3.0.0        | ❓    | ❓    | ❌     | ✅            | ✅           |
| 3.0.1        |       | ❓    | ❌     | ✅            | ✅           |
| 3.1.0        |       |       | ✅     | ❌            | ❌           |
| 3.1-SNAPSHOT |       |       |        | ✅            | ✅           |    
| 3.2-SNAPSHOT |       |       |        |               | ✅           |

### CODE REVIEW

**Review by PatrickRen — COMMENTED** [review]
@yuxiqian Thanks for the PR! The patch and test themselves looks good to me, but I'm a bit concerning about introducing new modules for each version, even patch versions   (like 3.0.1). As the project evolves there must be a lot of versions in the future, so number of modules could grow very fast. Not sure if this might burden the Maven build process.

**Comment by yuxiqian** [comment]
Thanks for @PatrickRen's comments! I too agree that it's not very elegant to package all history versions just for compatibility test, but seems to be inevitable since there's no such thing like C's macro or conditional compilation in Java.

What I'm thinking of is 1) moving those dummy projects to somewhere outside (like `yuxiqian/dummy-cdc-migration-test` repo), or 2) write a script to dynamically generate dummy projects during the CI. WDYT?

File: flink-cdc-migration-tests/pom.xml [hunk 1]
```diff
@@ -19,13 +19,10 @@ limitations under the License.
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
-    <parent>
```
Comment (PatrickRen) [inline flink-cdc-migration-tests/pom.xml]: Actually you can keep the <parent> section here. Removing `flink-cdc-migration-tests` from the <modules> section in the parent pom should be enough.

File: pom.xml [hunk 2]
```diff
@@ -40,7 +40,6 @@ limitations under the License.
         <module>flink-cdc-connect</module>
         <module>flink-cdc-runtime</module>
         <module>flink-cdc-e2e-tests</module>
-        <module>flink-cdc-migration-tests</module>
```
Comment (PatrickRen) [inline pom.xml]: Is it possible to use profiles to control whether to build `flink-cdc-migration-tests` module?

**Review by PatrickRen — COMMENTED** [review]
@yuxiqian Thanks for the update! I left some comments.

**Comment by yuxiqian** [comment]
Thanks for the tips! Addressed your comments to simplify pom files.

**Review by PatrickRen — APPROVED** [review]
@yuxiqian Thanks for the update! LGTM

**Pull request merged**